p2p/starknet/block_bodies.go

```diff
@@ -0,0 +1,164 @@
+package starknet
+
+import (
+	"slices"
+
+	"github.com/NethermindEth/juno/adapters/core2p2p"
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/p2p/starknet/spec"
+	"github.com/NethermindEth/juno/utils"
+	"google.golang.org/protobuf/proto"
+)
+
+type streamState int
+
+const (
+	_ streamState = iota
+
+	sendDiff // initial
+	sendClasses
+	sendBlockFin
+	terminal // final state
+)
+
+type blockBodyIterator struct {
+	bcReader blockchain.Reader
+	log      utils.Logger
+
+	state       streamState
+	header      *core.Header
+	stateUpdate *core.StateUpdate
+}
+
+func newBlockBodyIterator(bcReader blockchain.Reader, header *core.Header, log utils.Logger) *blockBodyIterator {
+	return &blockBodyIterator{
+		state:       sendDiff,
+		header:      header,
+		bcReader:    bcReader,
+		log:         log,
+		stateUpdate: nil, // to be filled during one of next states
+	}
+}
+
+func (b *blockBodyIterator) hasNext() bool {
+	statesHaveTransitions := []streamState{
+		sendDiff,
+		sendClasses,
+		sendBlockFin,
+	}
+	return slices.Contains(statesHaveTransitions, b.state)
+}
+
+// Either BlockBodiesResponse_Diff, *_Classes, *_Proof, *_Fin
+func (b *blockBodyIterator) next() (msg proto.Message, valid bool) {
+	switch b.state {
+	case sendDiff:
+		msg, valid = b.diff()
+		b.state = sendClasses
+	case sendClasses:
+		msg, valid = b.classes()
+		b.state = sendBlockFin
+	case sendBlockFin:
+		msg, valid = b.fin()
+		b.state = terminal
```

Comment:
I think this can be refactored to use `finAndSetTerminalState()` as it is doing similar things. It just seems strange that we have fin() which doesn't set the state to terminal, but then after its usage it sets the state to terminal immediately afterwards.

IMO we should only have one `fin()` and based on the current usages of the 2 fin functions i think it is possible.